Python Command Line Changes

ankitasanil opened the pull request:
These changes are relevant to the issue `#`1392 (https://github.com/MRtrix3/mrtrix3/issues/1392) and include below modifications:

- Creation of custom callable classes for handling different datatypes for positional and optional arguments on the command line. Used ArgumentTypeError for exception management instead of ValueError since the former allows allows adding custom error messages.
- Addition of newly defined callables (in app.py) for each positional and optional command line argument of mrtrix_cleanup and dwicat

@Lestropie - Please review these changes

ankitasanil committed:
Created custom callables for different datatypes on cmd

Created custom callable classes for handling different datatypes for positional and optional arguments on the command line. Used ArgumentTypeError for exception management instead of ValueError since the former allows allows adding custom error messages.

ankitasanil committed:
Used callable types in commands - mrtrix_cleanup and dwicat

Added newly defined callables (in app.py) for each positional and optional command line argument of mrtrix_cleanup and dwicat

Lestropie commented:
Is the use syntax definitely "`type=app.Parser().TypeInputImage()`" as opposed to "`type=app.Parser.TypeInputImage`"? I'd have expected the latter, in that it would be defining the type of the callable rather than an instance thereof.

Lestropie commented on lib/mrtrix3/app.py:
```diff
@@ -1101,6 +1101,88 @@ def _is_option_group(self, group):
            not group == self._positionals and \
            group.title not in ( 'options', 'optional arguments' )
 
+  class TypeBoolean:
+    def __call__(self, input_value):
+      true_var = "True"
+      false_var = "False"
+      converted_value = None
+      if input_value.lower() == true_var.lower():
+        converted_value = True
+      elif input_value.lower() == false_var.lower():
+        converted_value = False
+      else:
+        raise argparse.ArgumentTypeError('Entered value is not of type boolean')
+      return converted_value
```
For consistency with C++ command-line behaviour, see https://github.com/MRtrix3/mrtrix3/blob/3.0.4/core/mrtrix.h#L285-L293.

ankitasanil replied:
Changes made in commit: https://github.com/ankitasanil/mrtrix3/pull/1/commits/8d91a38c27bb2340584b4c3ecb5d3c9556dd6096

Lestropie commented on lib/mrtrix3/app.py:
```diff
@@ -1101,6 +1101,88 @@ def _is_option_group(self, group):
+
+  class TypeIntegerSequence:
+    def __call__(self, input_value):
+      seq_elements = input_value.split(',')
+      int_list = []
+      for i in seq_elements:
+        try:
+          converted_value = int(i)
+          int_list.append(converted_value)
+        except:
+          raise argparse.ArgumentTypeError('Entered value is not an integer sequence')
```
Good opportunity to use Python list comprehension.

Lestropie replied:
Also do not want to use all-encompassing exception catches; ideally only catch the type of exception that you *a priori* expect to be thrown if the user specifies something that isn't compatible with the expected type.

ankitasanil replied:
Changes made in commit: https://github.com/ankitasanil/mrtrix3/pull/1/commits/54bca482ace8b52acd93b374b184cbd99244cbb9

Lestropie commented on lib/mrtrix3/app.py:
```diff
@@ -1101,6 +1101,88 @@ def _is_option_group(self, group):
+      else:
+        return input_value
+
+  class TypeOutputDirectory:
+    def __call__(self, input_value):
+      return input_value
```
For a number of output types, check:
https://github.com/MRtrix3/mrtrix3/blob/3.0.4/core/app.cpp#L1181-L1235
From memory, when the `-force` option is used, there's a global "output files will be overwritten", but I think that in some instances there can be individual error messages of "this specific file / directory already exists, and will be overwritten at command completion"

tclose replied:
You mean a warning message needs to be added? The value of `--force` isn't going to be accessible to the output directory type is it?

Lestropie replied:
1.   Sorry yes, individual warning messages for individual items, not errors.
2.   In some cases we do issue an error; specifically attempting to overwrite a pre-existing fixel directory using `-force`. Can lead to unexpected behaviour, so we forbid it.
3.   Regarding knowledge of the force-overwrite status, this I don't know. It depends on the internals of `argparse`. But normally I'd read from `app.FORCE_OVERWRITE` in that case, which only gets set after `argparse.parse_args()` is finished, which means it probably will never be set when those callables are invoked.
     Better, which bears some relation to #1975, would be for the callable to yield some class for which a pre-existing file at that path could be checked at the first instance at which the executing script reads from that path.
4.   Pretty sure that the situation where individual warning messages for overwriting individual files is specific to the Python scripts and doesn't happen in the C++ binaries. So if that no longer happens as a natural consequence of these changes, it's not the end of the world, especially given that up to this point it had to be hard-coded on a per-case basis.

tclose replied:
Yeah, you won't be able to get it to work from inside the type callable. You would need to add some code to the app I think to loop through all types and find the file/directory types

Lestropie commented on lib/mrtrix3/app.py:
```diff
@@ -1101,6 +1101,88 @@ def _is_option_group(self, group):
+    def __call__(self, input_value):
+      return input_value
+
+  class TypeInputTractogram:
+    def __call__(self, input_value):
+      if not input_value.endsWith('.tck'):
+        raise argparse.ArgumentTypeError(input_value + ' is not a valid track file')
+      return input_value 
```
Tractograms may be an opportunity to use some inheritance. Any checks that apply to input / output files additionally apply to tractograms, but then checking the file extension is an *additional* task that applies to tractogram files but not generic files.

ankitasanil replied:
Changes made in commit: https://github.com/ankitasanil/mrtrix3/pull/1/commits/8d91a38c27bb2340584b4c3ecb5d3c9556dd6096

Lestropie replied:
That commit duplicates within `TypeInputTractogram` the checks previously performed only in `TypeInputFile`. However I was indicating that that should be achieved through class inheritance.

ankitasanil replied:
Changes made in commit: https://github.com/ankitasanil/mrtrix3/pull/1/commits/6dcfe5353c171a7a86aa2109b175f667aecf1fb2

ankitasanil committed:
Changes done as instructed in the review

Updated the logic for TypeBoolean class for consistency with C++ command-line behaviour
Added new checks in TypeInputTractogram class for file validation

ankitasanil committed:
Used callable types in few more commands

Added callables for each positional and optional command line argument in dwifslpreproc, dwigradcheck, dwishellmath, labelsgmfix, mask2glass, population_template and responsemean

Lestropie commented:
Preferable for class names to echo those used in C++ code:
https://github.com/MRtrix3/mrtrix3/blob/3.0.4/core/cmdline_option.h#L44-L62

Lestropie committed:
Python API: Integrate argument types across commands

Co-authored-by: Ankita Sanil <ankitasanil@gmail.com>

ankitasanil committed:
Changed the syntax as per https://github.com/ankitasanil/mrtrix3/pull/1#issuecomment-1418494241

Used the new syntax as "type=app.Parser.TypeInputImage()" across all Python API commands

ankitasanil committed:
Python API: Used Python list comprehension as per https://github.com/ankitasanil/mrtrix3/pull/1#discussion_r1096932040

Replaced the traditional for loop with list comprehension in TypeIntegerSequence and TypeFloatSequence classes

Lestropie committed:
population_template: Refine cmdline interface

Lestropie commented:
While in cd33930 for `population_template` I've made modifications to the loading of data from command-line options, these are not tested as part of CI. We would need to augment the script test data with multi-contrast data from multiple subjects, and add tests that use all possible command-line options.

Lestropie committed:
Change interfaces for robust linear registration estimators

Applies to both population-template and mrregister.
Makes "none" a valid selection of robust estimator in both cases.

ankitasanil committed:
Python API: Updated class names as per https://github.com/ankitasanil/mrtrix3/pull/1#issuecomment-1420052303

Updated class names across all commands to be in sync with C++ code

ankitasanil committed:
Modifications in existing code to handle new argument types

Lestropie commented:
From a few different sources, it looks like custom Actions should derive from `argparse.Action`.

ankitasanil committed:
Modifications in existing code to handle new argument types (contd.)

ankitasanil committed:
Added inheritance for TracksIn checks as per (https://github.com/ankitasanil/mrtrix3/pull/1#discussion_r1099607175)

Implemented class inheritance to avoid duplicate checks for tractogram input files. Instead, reused the checks from ArgFileIn type via inheritance.

ankitasanil committed:
Python API: Ability to handle piped commands

Changes for handling piped images in the Python API scripts. However, this implementation does not include deletion of the temp/piped images at the end of the command execution.

ankitasanil committed:
Changes for output image piping

The current implementation is temporary since it doesn't cover all the use-cases. However, it supports a working scenario.

tclose commented:
> Is the use syntax definitely "`type=app.Parser().TypeInputImage()`" as opposed to "`type=app.Parser.TypeInputImage`"? I'd have expected the latter, in that it would be defining the type of the callable rather than an instance thereof.

Looks ok to me. `app.Parser.Type*` need to be callable objects so they can return an object of different type, i.e. a `bool` instead of `app.Parser.Boolean`

tclose commented:
@Lestropie , just trying to work out what is necessary to finish off this PR. Is it just a few stylistic changes you want first?

tclose commented:
> Preferable for class names to echo those used in C++ code: https://github.com/MRtrix3/mrtrix3/blob/3.0.4/core/cmdline_option.h#L44-L62

Could they even mirror them more closely by using snake_case, e.g. `type_image_in()` (would still be a class under the hood but no one needs to know that), or perhaps just `image_in()` since it is going to be set to type, i.e. `type=image_in()`

tclose commented:
> From a few different sources, it looks like custom Actions should derive from `argparse.Action`.

which are the custom Actions?

Lestropie commented:
> what is necessary to finish off this PR

-   Hoped for confirmation from @ankitasanil regarding whether she wanted to contribute to finishing the feature set.
    (If not, will pull this branch into the main repo and make a PR there)
-   Anything outstanding in #1392, particularly piping of image data given that was the primary goal
-   Contemplation of class naming / placement. Been a while since I looked at / contemplated it, but I would probably opt for something that complies with convention / `pylint` enforcement for class names but has word ordering consistent with the corresponding C++ modifier functions.

> which are the custom Actions?

This might be a red herring on my part. I may have temporarily mixed up types and actions. Though it also depends on the most appropriate way to handle image piping.

tclose commented:
> > what is necessary to finish off this PR
> 
> * Hoped for confirmation from @ankitasanil regarding whether she wanted to contribute to finishing the feature set.
>   (If not, will pull this branch into the main repo and make a PR there)
> * Anything outstanding in [`run.command()`: Support for command-line piping MRtrix3/mrtrix3#1392](https://github.com/MRtrix3/mrtrix3/issues/1392), particularly piping of image data given that was the primary goal

Would it be possible to split them up if there are other issues in #1392 that prove tricky?

> * Contemplation of class naming / placement. Been a while since I looked at / contemplated it, but I would probably opt for something that complies with convention / `pylint` enforcement for class names but has word ordering consistent with the corresponding C++ modifier functions.

Personally I find pylint pretty annoying (flake8 is more reasonable IMO), and in such cases where the user doesn't need to know whether it is a class or function there isn't much benefit trying to stick to the convention.

> > which are the custom Actions?
> 
> This might be a red herring on my part. I may have temporarily mixed up types and actions. Though it also depends on the most appropriate way to handle image piping.

Yeah, I think actions are for different use cases

Lestropie commented:
> Would it be possible to split them up if there are other issues in https://github.com/MRtrix3/mrtrix3/issues/1392 that prove tricky?

Definitely don't want to skip over the piping; appreciate that in your scenario you're invested in typed arguments, but from our perspective that was actually just a means to an end, so stopping short of that mark would be kind of unusual. And wouldn't want to split for the sake of having typed arguments for Pydra if it then turned out that we need to make further backward-incompatible back-end changes to facilitate the piping.

The other stuff should not be terribly difficult. Using `argparse.FileType` I'm happy to skip, but we do need basic sanity check tests in place, and that should be easy as we shouldn't need to modify the test data repo.

tclose commented:
> > Would it be possible to split them up if there are other issues in [MRtrix3#1392](https://github.com/MRtrix3/mrtrix3/issues/1392) that prove tricky?
> 
> Definitely don't want to skip over the piping; appreciate that in your scenario you're invested in typed arguments, but from our perspective that was actually just a means to an end, so stopping short of that mark would be kind of unusual. And wouldn't want to split for the sake of having typed arguments for Pydra if it then turned out that we need to make further backward-incompatible back-end changes to facilitate the piping.
> 
> The other stuff should not be terribly difficult. Using `argparse.FileType` I'm happy to skip, but we do need basic sanity check tests in place, and that should be easy as we shouldn't need to modify the test data repo.

ok, fair enough. Let me know if there is anything I can do to push it along

Lestropie commented:
Given you're active and eager, I'll pull these commits over to a branch on the main repo, try to make the primary up-front changes I'd want to make, and then we can see what's left from there.
